docs/examples/example_survival.ipynb

```diff
@@ -0,0 +1,206 @@
+    "        y_lower_bound = np.array(y)\n",
+    "\n",
+    "        dtrain.set_float_info(\"label_lower_bound\", y_lower_bound)\n",
+    "        dtrain.set_float_info(\"label_upper_bound\", y_upper_bound)\n",
+    "\n",
+    "        if self.booster_params is None:\n",
+    "            booster_params = common_params\n",
+    "        else:\n",
+    "            booster_params = common_params | self.booster_params\n",
+    "\n",
+    "        self.bst_ = xgb.train(booster_params, dtrain, num_boost_round=self.n_estimators)\n",
+    "        return self\n",
+    "\n",
+    "    def predict(self, X):\n",
+    "        # Check if fit has been called\n",
+    "        check_is_fitted(self)\n",
+    "        # X contains \"censored\" column\n",
+    "        features = set(X.columns) - {\"censored\"}\n",
+    "        if set(np.unique(X[\"censored\"])) != {0, 1}:\n",
+    "            raise ValueError(\"censored column should be binary.\")\n",
+    "        dtest = xgb.DMatrix(X[list(features)], enable_categorical=True)\n",
+    "        return self.bst_.predict(dtest)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Fitting the MetaLearner\n",
+    "\n",
+    "Finally we can instantiate any MetaLearner with the ``SurvivalModel`` as the nuisance model and estimate the CATEs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from lightgbm import LGBMClassifier, LGBMRegressor\n",
+    "from metalearners import TLearner, DRLearner, RLearner\n",
+    "from metalearners.utils import simplify_output\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "tlearner = TLearner(False, 2, SurvivalModel, nuisance_model_params={\"n_estimators\": 5})\n",
+    "tlearner.fit(X, y, w)\n",
+    "cate_estimates_tlearner = simplify_output(tlearner.predict(X, True))\n",
+    "\n",
+    "drlearner = DRLearner(\n",
+    "    False,\n",
+    "    2,\n",
+    "    nuisance_model_factory=SurvivalModel,\n",
+    "    propensity_model_factory=LGBMClassifier,\n",
+    "    treatment_model_factory=LGBMRegressor,\n",
+    "    nuisance_model_params={\"n_estimators\": 5},\n",
+    "    propensity_model_params={\"verbose\": -1, \"n_estimators\": 5},\n",
+    "    treatment_model_params={\"verbose\": -1, \"n_estimators\": 5},\n",
+    ")\n",
+    "drlearner.fit(X, y, w)\n",
+    "cate_estimates_drlearner = simplify_output(drlearner.predict(X, True))\n",
+    "\n",
+    "rlearner = RLearner(\n",
+    "    False,\n",
+    "    2,\n",
+    "    nuisance_model_factory=SurvivalModel,\n",
+    "    propensity_model_factory=LGBMClassifier,\n",
+    "    treatment_model_factory=LGBMRegressor,\n",
+    "    nuisance_model_params={\"n_estimators\": 5},\n",
+    "    propensity_model_params={\"verbose\": -1, \"n_estimators\": 5},\n",
+    "    treatment_model_params={\"verbose\": -1, \"n_estimators\": 5},\n",
+    ")\n",
+    "rlearner.fit(X, y, w)\n",
```

Comment:
I think for the T-learner you can indeed use all the observations, but in the treatment model of the R-learner for example, you should only use the non-censored observations with Inverse Probability of Censoring Weighting (see Equation 22 of this paper for [more details](https://arxiv.org/pdf/2207.07758)). Would that be compatible with the `metalearners` library? What do you think @FrancescMartiEscofetQC @kklein ?

Reply:
Using different observations for the different models is not possible with the current implementation. I think it would introduce some problems because we rely on the fact that all models have been trained on the same dataset as all models are wrapped with `CrossFitEstimator`. I believe in the particular case of the RLearner it could be adapted as the second stage model is not synchronized with the first stage models so a subselection could be done but I doubt a generalization could work for all metalearners (although not sure if this has to be done in all the metalearners).

On the other side, using weights for the `RLearner` is not possible right now because of [this lines](https://github.com/Quantco/metalearners/blob/60745dc284980f60f129cbef443570bd2acd6b5b/metalearners/rlearner.py#L120-L127), but I think this could be adapted easily and allow passing sample weights for the treatment model and take the product of the provided by the user and the ones computed in the metalearner.

Reply:
I second @FrancescMartiEscofetQC's view that truly removing observations for some base models will not fare well with the general paradigm since the cross-fitting of approaches of individual base models rely on each other. If they face different quantities of data, this will fail.

What we should be able to do, though, would be to set the weights of certain observations for certain base models to 0 (as long as the base model of choice is fine with that).

As @FrancescMartiEscofetQC pointed out, this is currently not possible for the R-Learner, but should be feasible for all other MetaLearners.

@FrancescMartiEscofetQC should we then simply use the DR-Learner in this example, train a censoring model and pass its predictions as weights to the DR-Learner treatment model? Would that seem appropriate to you @AronPapQC ?

Reply:
I see, thank you very much for the explanation @FrancescMartiEscofetQC ! 

> @FrancescMartiEscofetQC should we then simply use the DR-Learner in this example, train a censoring model and pass its predictions as weights to the DR-Learner treatment model? Would that seem appropriate to you @AronPapQC ?

Sounds good to me, maybe I would just also add a comment in this example notebook that the indirect metalearners are already compatible with time-to-event outcomes, however not all direct metalearners (e.g. R-learner) are compatible yet because of the subsampling for the treatment model.

Reply:
> What we should be able to do, though, would be to set the weights of certain observations for certain base models to 0 (as long as the base model of choice is fine with that).

That's a good idea, doing some tests looks like it only checks for non-negativity, so giving a weight of 0 looks possible.

Regarding the censoring model I see how we could do it but I have doubts regarding the correctness in our case. We are doing it with nuisance models which predict the time-to-event but in the paper they work with nuisances that predict the risk (the probability that someone survives past a fixed $t_0$), do you have any reference which would provide support for doing this @AronPapQC ? Or how did you solve it in your team?

Reply:
Kind bump @AronPapQC :)

Reply:
> We are doing it with nuisance models which predict the time-to-event but in the paper they work with nuisances that predict the risk (the probability that someone survives past a fixed 
), do you have any reference which would provide support for doing this @AronPapQC ? Or how did you solve it in your team?

We mainly have models that predict the risk & survival curves (e.g. Cox-PH) and not the time-to-event directly, but I think the censoring weighting would still be correct in case of only using the complete (e.g. uncensored) observations in the CATE model, see section 3 of [this paper](https://www.jstor.org/stable/2673467) for motivation/reference:

<img width="984" alt="image" src="https://github.com/Quantco/metalearners/assets/113583019/bce18aac-8b1f-43a8-9a8e-7ca390a33a3c">

Let me know whether that makes sense @FrancescMartiEscofetQC ?

Reply:
> Edit: I'm thinking that I did it wrong and I should put a 0 weight to the censored units, am I right @AronPapQC ?

In the treatment model yes!